### DISCUSSION TITLE
SSR Oauth with next js pages router

### DISCUSSION THREAD

**Discussion by alecswjo** [supabase/supabase#20911] [Questions]
# Chore
SSR Oauth with next js pages router

## Describe the chore
SSR Oauth doesn't work with next js pages router. The example here (https://supabase.com/docs/guides/auth/server-side/oauth-with-pkce-flow-for-ssr?framework=nextjs) uses apps router. 

Since { cookies } from 'next/headers' needs a server component (not in pages router). i'm not sure how to implement. Does anyone have any tips?

**Comment by charislam** [comment]
The flow is very similar, just that you need to get your cookies from the request object instead of the `cookies` function. (I actually prefer that pattern: aligns better with the platform, and less magic.)

Check out the Next.js Starter Guide for how to set up a Supabase client inside your API route that reads the cookies from the request: https://supabase.com/docs/guides/auth/server-side/nextjs?router=pages

Then it's just a matter of substituting the client. Everything else should be relatively router-agnostic (barring some minor differences in the request object and redirect function, which you can check the [Next.js docs](https://nextjs.org/docs/pages/api-reference/functions/next-request) for).

**Reply by big-nico** [reply]
Appreciate your response. 

After creating the Supabase client (following the Next.js Starter Guide from Supabase) and redirecting to the home page, the `user` object continues to be undefined. 

Here is my `getServerSideProps` which checks that the user is defined on protected pages: 

```
export async function getServerSideProps(context: GetServerSidePropsContext) {
  const supabase = createClient(context);
  const { data, error } = await supabase.auth.getUser();

  if (error || !data) {
    return {
      redirect: {
        destination: '/',
        permanent: false,
      },
    };
  }

  return {
    props: {
      user: data.user,
    },
  };
}
```

Is the pages router fully supported with the new supabase ssr package?

**Reply by charislam** [reply]
Yep that guide was recently tested with the Pages Router.

I'd need to see a bigger repro to determine where the cookie is failing to be set, any chance you could link me to an example repo?

**Reply by big-nico** [reply]
https://github.com/big-nico/nextjs-supabase-pages-starter

Here is a sample repo - all you have to do is make sure the supabase project you connect has google oauth enabled.

**Reply by charislam** [reply]
Ahhh you know what, you're right, it's our problem. In the cookies methods for the API handler, it should be `appendHeader`, not `setHeader`. We must have missed that one in our testing, I'll put in a PR to fix it.

Thanks for the catch!

**Reply by big-nico** [reply]
Exciting stuff! Let me know when the PR is out & how to incorporate the latest into our project.

**Reply by charislam** [reply]
Hey @big-nico, no need to wait on us! The instructions are the problem, not the package, so in your `utils/supabase/api.ts`, just swap out `res.setHeader` for `res.appendHeader`!

**Reply by big-nico** [reply]
lovely!

**Comment by riccardolardi** [comment]
Hi, we're in the same situation, still on legacy pages router but migrating to supabase/ssr package. Which client to use in `middleware.ts` though?

**Comment by Stephcraft** [comment]
The latest version of the Next.js Pages router `supabase/util` scripts can be found here:

[`server-props.ts`](https://github.com/supabase/supabase/blob/7d722bfc0360c564cebf27eb5253d20ba09c1270/apps/docs/content/guides/auth/server-side/nextjs.mdx?plain=1#L638) for `getServerSideProps()` — *The new App router [`server.ts`](https://github.com/supabase/supabase/blob/1cda6aaa10a13d849fc323b69e5b83b27ac74e8d/examples/auth/nextjs/utils/supabase/server.ts) does not work on the Page router*
```ts
import { type GetServerSidePropsContext } from 'next'
import { createServerClient, serializeCookieHeader } from '@supabase/ssr'

export function createClient({ req, res }: GetServerSidePropsContext) {
    const supabase = createServerClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY!,
        {
            cookies: {
                getAll() {
                    return Object.keys(req.cookies).map((name) => ({ name, value: req.cookies[name] || '' }))
                },
                setAll(cookiesToSet) {
                    res.setHeader(
                        'Set-Cookie',
                        cookiesToSet.map(({ name, value, options }) =>
                            serializeCookieHeader(name, value, options)
                        )
                    )
                }
            }
        }
    )

    return supabase
}
```

[`static-props.ts`](https://github.com/supabase/supabase/blob/7d722bfc0360c564cebf27eb5253d20ba09c1270/apps/docs/content/guides/auth/server-side/nextjs.mdx?plain=1#L667) for `getStaticProps()` — *Can be replaced by the [admin client](https://supabase.com/docs/reference/javascript/admin-api)*
```ts
import { createClient as createClientPrimitive } from '@supabase/supabase-js'

export function createClient() {
    const supabase = createClientPrimitive(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY!
    )

    return supabase
}
```

[`components.ts`](https://github.com/supabase/supabase/blob/7d722bfc0360c564cebf27eb5253d20ba09c1270/apps/docs/content/guides/auth/server-side/nextjs.mdx?plain=1#L680) for React client components — *Same as the new [`client.ts`](https://github.com/supabase/supabase/blob/1cda6aaa10a13d849fc323b69e5b83b27ac74e8d/examples/auth/nextjs/utils/supabase/client.ts)*
```ts
import { createBrowserClient } from '@supabase/ssr'

export function createClient() {
  const supabase = createBrowserClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY!
  )

  return supabase
}
```

[`api.ts`](https://github.com/supabase/supabase/blob/7d722bfc0360c564cebf27eb5253d20ba09c1270/apps/docs/content/guides/auth/server-side/nextjs.mdx?plain=1#L693) for API routes `pages/api/...` — *Same as `server-props.ts`, except for function signature parameters*
```ts
import { createServerClient, serializeCookieHeader } from '@supabase/ssr'
import { type NextApiRequest, type NextApiResponse } from 'next'

export default function createClient(req: NextApiRequest, res: NextApiResponse) {
    const supabase = createServerClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY!,
        {
            cookies: {
                getAll() {
                    return Object.keys(req.cookies).map((name) => ({ name, value: req.cookies[name] || '' }))
                },
                setAll(cookiesToSet) {
                    res.setHeader(
                        'Set-Cookie',
                        cookiesToSet.map(({ name, value, options }) =>
                            serializeCookieHeader(name, value, options)
                        )
                    )
                }
            }
        }
    )

    return supabase
}
```

[`middleware.ts`/`proxy.ts`](https://supabase.com/docs/guides/auth/server-side/creating-a-client?queryGroups=framework&framework=nextjs#hook-up-middleware) continues to be compatible on both the App and Pages router.

These are from the last commit ([`7d722bf`](https://github.com/supabase/supabase/commit/7d722bfc0360c564cebf27eb5253d20ba09c1270)  *Nov 2025*) modifying the legacy Next.js documentation right before the commit deleting the page ([`534c300`](https://github.com/supabase/supabase/commit/534c300a7adb537e58eb961fd1ccf986b62a3498#diff-35dd19c72705f71804899e011ba6f9e48ab100743dc8e10ee21877423f9e04f8)). The history of the commits for this documentation can be found [here](https://github.com/supabase/supabase/commits/master/apps/docs/content/guides/auth/server-side). Idk where the actual example files for these code snippets ([`examples/auth/nextjs/utils/supabase/`](https://github.com/supabase/supabase/tree/master/examples/auth/nextjs/utils/supabase)) are though.
